### ISSUE TITLE
Swatches no longer crossed out after indexes invalidated and reindexed

### ISSUE THREAD

**Issue by dphilipps** [magento/magento2#34127]
### Preconditions (*)

1. Magento 2.4.3-develop 
2. Issue also exists in official 2.4.3 release

### Steps to reproduce (*)

1. In admin, go to System->Index Management
2. Select all indexes, followed by Actions->Update by Schedule->Submit
3. On front end, go to product /juno-jacket.html
4. Add 100 of XL/Green to basket and proceed to checkout and place order to ensure stock becomes fully sold out
5. Visit product page again and click on XL option
6. Green swatch is correctly crossed-out (if not, repeat from step 4 with a different option -see comments below)
7. In admin, go to System->Index Management
8. Take note of Updated time stamp
9. Select all indexes, followed by Actions->Invalidate Index->Submit
10. Refresh as necessary and wait for updated time stamp to change (compared to point 8 above) meaning indexes have reindexed
11. On front end, go to product /juno-jacket.html and click on XL option
12. Green swatch is no longer crossed out

### Expected result (*)

1. The green swatch should still be crossed out after the index has been invalidated and reindexed
![image](https://user-images.githubusercontent.com/44231639/133994729-851b3977-939a-43a2-a0c6-40f3089bad56.png)

### Actual result (*)

1. The green swatch IS NOT crossed out after the index has been invalidated and reindexed.  The unsalable product can be selected and 'Add to Cart' button pressed.  However 'The requested qty is not available' message is displayed after adding to cart
![image](https://user-images.githubusercontent.com/44231639/133994785-1294b79e-a249-4f21-be40-81138196ed2c.png)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by dphilipps** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
Occasionally the sold-out product is not initially crossed-out after placing an order as expected.  There is another ticket open which covers this #33799 however I assume it is actually a manifestation of the issue described in this ticket and caused by indexing and the inconsistency caused by the timing of the index update after purchase (only a guess though).

**Comment by dphilipps** [comment]
I have verified the reproduction steps with the above instance.

**Comment by dphilipps** [comment]
Issue: Clear Description

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the behavior is different:

Steps performed:

1. Created a configurable product with options. (used color attribute - visual swatch)
2. Front-end - Selected one option with all quantity and placed order - No issue
3. Front-end - Product Details page - Red Swatch is not crossed out
4. Click on Red Swatch and Add to Cart
Error: "The requested qty is not available" error message is displayed to the user.
5. In admin, go to System->Index Management
Select all indexes, followed by Actions->Invalidate Index->Submit
6. Reindex and clear cache from CLI
php bin/magento indexer:reindex
php bin/magento cache:flush
7. Front-end - Product Details page - Red Swatch is not crossed out
8. Click on Red Swatch and Add to Cart
Error: "The requested qty is not available" error message is displayed to the user.
User is not able to place order before or after reindexing.
@dphilipps , can you please recheck the issue on Magento 2.4-develop branch and provide missing steps if any if the issue is still reproducible.
Thank you.
![image](https://user-images.githubusercontent.com/60198231/134510997-1dee0a0e-60b9-4ec2-ab32-8b8c81a9a562.png)

**Comment by dphilipps** [comment]
It looks as though you didn't complete steps 1 & 2 according to your write up?
1) In admin, go to System->Index Management
2) Select all indexes, followed by Actions->Update by Schedule->Submit

The challenge with this issue is that there seems to possibly be a different bug which is preventing the swatch being crossed out initially (after purchasing all).  This bug (the state of the swatch reverting) obviously cannot be reproduced without the swatch first being crossed out.

Can I kindly ask, when after the indexers are changed to scheduled (1 & 2 above), if a swatch is not crossed out, please purchase another option (buying all stock, typically 100 items) and try again.  If often works on subsequent attempts.

Once the crossed out swatch has been obtained, then my steps 7 onwards can be followed to reproduce the issue.

**Comment by dphilipps** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
Hi @engcom-November 
To confirm, I have followed my reproduction steps again on a new instance at at step 6, I needed to repeat the purchased for a second option.  After the second purchase (same colour in different size) but swatches became crossed out correctly.  I then repeated a third purchase which worked as well.

Following the steps from 7 then reproduced the issue.

PLEASE NOTE After the indexes have been invalidated, you must wait for cron to run again and reindex before going back to check the status of the swatch.

I note  you followed slightly different steps to me.  I recommend following my steps exactly and purchasing the demo products, this will also save time over creating a configurable.

Many thanks

**Comment by engcom-November** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible with below Steps performed:
**Steps to reproduce:**
Create a configurable product with options. (used color attribute - visual swatch) / Or Use sample data  /juno-jacket.html which has options.
In admin, go to System->Index Management
Select all indexes, followed by Actions->Update by Schedule->Submit. (Make sure cron jobs are running)
Front-end: Add qty - 100 of XL/Green to basket and proceed to checkout and place order to ensure stock becomes fully sold out
Visit product page again and click on XL option
Reindex and Clear cache
**Issue 1:** Green swatch is not correctly crossed-out
Select Same color again and different size ex: M and Place Order with all quantity - 100
Go to Product Details page and select XL or M
**There is no issue now - Green Swatch is crossed out now.** 
System->Index Management - Select All -  Actions-> Invalidate Index
Wait till all indexers are reindexed
Refresh Product Details page - @dphilipps . No issue. Green Swatch is correctly crossed out even after reindexing.

Confirming this issue for Issue 1 mentioned here and inconsistent behavior while adding out of stock product to cart if the swatch is not crossed out.
![image](https://user-images.githubusercontent.com/60198231/134917333-0026b615-dd0f-4935-b97e-92352d4cd554.png)
![image](https://user-images.githubusercontent.com/60198231/134917393-048d746a-8926-4451-81d3-d26fdb00b3c4.png)
For newly created products - Error: "The requested qty is not available" error message is displayed to the user.

**Comment by dphilipps** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
Hi @engcom-November
Thanks for trying again!  I notice in your reproduction steps, 'submit' is missing after invalidating the indexes:

System->Index Management - Select All - Actions-> Invalidate Index

This might be a silly question, but did you definitely submit after selecting Invalidate Index in the dropdown?

I've just followed your steps again without 'Reindex and Clear cache' as this isn't in my repro steps; on a new instance (above) and the swatches are no longer crossed out after the indexes are invalidated and reindexed!  

I'm perplexed why I can reproduce this each time and you are unable.  Would a video help?

**Comment by engcom-November** [comment]
Hi @dphilipps ,
Yes, I did submit after invalidating index
Please find the reference video of the same for your understanding.
/home/global/Downloads/swatch_test.mp4
Also, Issue 1 mentioned above is also not reproducible now after reindexing and clearing cache (Index Management - Update by Schedule - Once all the indexers are updated - Swatches are crossed out. We dont need to make another attempt of buying all quantity with other option)
Can you please recheck the issue on Magento 2.4-develop branch and ensure rechecking only after all Indexers are updates successfully.

**Comment by dphilipps** [comment]
Hi
How do I access the video ?

Thanks
Dave

**Comment by engcom-November** [comment]
Hi @dphilipps ,
Yes, I did submit after invalidating index
Please find the reference video of the same for your understanding.
https://user-images.githubusercontent.com/60198231/135072492-470bca71-d753-4ce1-9110-69cb99a6eb96.mp4

Also, Issue 1 mentioned above is also not reproducible now after reindexing and clearing cache (Index Management - Update by Schedule - Once all the indexers are updated - Swatches are crossed out. We dont need to make another attempt of buying all quantity with other option)
Can you please recheck the issue on Magento 2.4-develop branch and ensure rechecking only after all Indexers are updates successfully.

**Comment by dphilipps** [comment]
Thanks for the video - you are following the right steps.  The one difference is the environment.  I see you are working on a local instance, but have you tried the steps using a 2.4-develop instance here?  That is what I am using to reproduce the problem and it works every time.  Could you try that and let me know how you get on please?

**Comment by dphilipps** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
Hi @engcom-November
Here is a video showing my reproduction steps on the above instance from the magento-deployment-service.  As you'll see the swatch is no longer crossed after the indexes are invalidated and rebuilt.

https://www.loom.com/share/70c277bc1521477cb09c0c91d5a68231

Please let me know if any questions

Thanks
Dave

**Comment by engcom-November** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
@engcom-Alfa 
Whilst sounding similar and having a similar effect, this is actually a **different issue with a different use-case**.

This issue occurs when **a manual reindex is done, or indexes invalidated via admin**, causing a reindex.  This reverts/resets the salabililty data which is set by the **inventory.reservations.updateSalabilityStatus** consumer when a product sells out.

The related bug https://github.com/magento/magento2/issues/33799 describes how the product is not being crossed out (becoming unsalable) immediately after placing an order when a product comes sold out.  This bug (34127) manifests itself when a product is already successfully crossed out, and is in fact a precondition of this bug.  

Please can you reopen this bug?

Let me know if you need any more info.
